### PR TITLE
feat(toolbar): toolbar will now persist selected dataSource and database across intellij sessions and will restore them on boot INTELLIJ-65

### PR DESCRIPTION
**Pull request by himanshusinghs**
## Description
This PR introduces a persistent state for Toolbar - selected DataSource and Database are persisted and restored for different IntelliJ project's sessions.

There are some important refactors alongside the main work, summarised below:
- EditorToolbarDecorator is refactored into a ProjectActivity to correctly initialise the toolbar only when the project is opened
  - This fixes multiple different initialisation of the toolbar, subscription set ups on every editor created lifecycle event
- introduces MdbPluginDisposable as a dedicated project service to act as a disposable parent for different subscription that we intialise during the plugin's operation. This service also exposes an interface to register cleanup functions for when a project is closed or plugin is unloaded.
  - This fixes the memory leak issue of the subscriptions that we were creating on editor created lifecycle event
- introduced Project level services - EditorService, DataSourceService to encapsulate logic for working with IntelliJ resources such as Editor, LocalDataSourceManager, etc
- introduced DataSourceModel and DatabaseModel to consolidate and encapsulate actions taken on DataSource/Database selection as it required interfacing with two different services
- refactored UI components - DataSourceComboBox, DatabaseComboBox and MdbJavaEditorToolbar to better encapsulate logic of working with internal state

### Checklist

- [ ] New tests and/or benchmarks are included.
- [ ] Documentation is changed or added.
- [ ] Changelog is updated accordingly.
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement).

### CODE REVIEW

File: packages/jetbrains-plugin/src/main/kotlin/com/mongodb/jbplugin/editor/EditorToolbarDecorator.kt [hunk 1]
```diff
@@ -1,233 +1,142 @@
...
-        (event.newEditor as? TextEditor)?.editor?.let { newEditor ->
-            editor = newEditor
-            editor.project?.let { project ->
-                guessAndStoreDialect(project)
-            }
-            ensureToolbarIsVisibleIfNecessary()
-            toolbar.reloadDatabases()
+    private fun runReadActionWhenInitialised(block: () -> Unit) {
+        ApplicationManager.getApplication().runReadAction {
+            block()
```
Comment (kmruiz) [inline packages/jetbrains-plugin/src/main/kotlin/com/mongodb/jbplugin/editor/EditorToolbarDecorator.kt:52]: Probably forgot to rename this function to something else :D

File: packages/jetbrains-plugin/src/main/kotlin/com/mongodb/jbplugin/editor/inputs/DatabaseModel.kt [hunk 2]
```diff
@@ -0,0 +1,53 @@
+package com.mongodb.jbplugin.editor.inputs
+
+import com.intellij.database.dataSource.LocalDataSource
+import com.mongodb.jbplugin.editor.DataSourceService
+import com.mongodb.jbplugin.editor.EditorService
+import com.mongodb.jbplugin.editor.DatabasesLoadingState
+import com.mongodb.jbplugin.editor.ToolbarSettings
+
+sealed interface ComboBoxStateLoadingState {
```
Comment (kmruiz) [inline packages/jetbrains-plugin/src/main/kotlin/com/mongodb/jbplugin/editor/inputs/DatabaseModel.kt:9]: That's a wonderful usage of a sealed interface.

File: packages/jetbrains-plugin/src/main/kotlin/com/mongodb/jbplugin/inspections/impl/FieldCheckInspectionBridge.kt [hunk 3]
```diff
@@ -177,7 +177,7 @@ internal object FieldCheckLinterInspection : MongoDbInspection {
             descriptor: ProblemDescriptor,
         ) {
             val editor = FileEditorManager.getInstance(project).selectedTextEditorWithRemotes[0] ?: return
-            MdbJavaEditorToolbar.showModalForSelection(editor, coroutineScope)
+            // MdbJavaEditorToolbar.showModalForSelection(editor, coroutineScope)
```
Comment (kmruiz) [inline packages/jetbrains-plugin/src/main/kotlin/com/mongodb/jbplugin/inspections/impl/FieldCheckInspectionBridge.kt:180]: I'm not sure this should be commented?

```suggestion
            MdbJavaEditorToolbar.showModalForSelection(editor, coroutineScope)
```

File: packages/jetbrains-plugin/src/main/resources/META-INF/plugin.xml [hunk 4]
```diff
@@ -34,7 +34,8 @@
 
     <extensions defaultExtensionNs="com.intellij">
         <postStartupActivity implementation="com.mongodb.jbplugin.ActivatePluginPostStartupActivity"/>
-        <editorFactoryListener implementation="com.mongodb.jbplugin.editor.EditorToolbarDecorator" />
+        <postStartupActivity implementation="com.mongodb.jbplugin.editor.EditorToolbarDecorator"/>
+<!--        <editorFactoryListener implementation="com.mongodb.jbplugin.editor.EditorToolbarDecorator" />-->
```
Comment (kmruiz) [inline packages/jetbrains-plugin/src/main/resources/META-INF/plugin.xml:38]: Let's just remove it 🔥 

```suggestion

```

**Review by kmruiz — APPROVED** [review]

File: packages/jetbrains-plugin/src/main/kotlin/com/mongodb/jbplugin/editor/EditorToolbarDecorator.kt (hunk 1 above)
Comment (himanshusinghs) [reply packages/jetbrains-plugin/src/main/kotlin/com/mongodb/jbplugin/editor/EditorToolbarDecorator.kt]: yea I did not include the logic to actually check wether the activity started. Have done that now 👍

**Pull request merged**